plugins/cr8-copy-s3/src/main/java/io/crate/copy/s3/S3FileInput.java

```diff
@@ -146,6 +152,38 @@ static URI toPreGlobUri(URI fileUri) {
         return preGlobUri;
     }
 
+    @VisibleForTesting
+    static URI reformat(URI uri) {
```

Comment:
```suggestion
    static URI normalize(URI uri) {
```

Reply:
Could merge this with the `BucketAndKey` logic and create a `S3Uri` class:

```
S3URI:

  static S3URI parse(URI uri) {...}

  normalizedURI :: URI
  bucket :: String
  key :: String
```

I think this would reflect the normalization of the URI a bit better in the type system.
The `S3FileInput` would then have the S3URI as property instead of the URI.